### ISSUE TITLE
types: definePlugin infers type incorrect when init is written in function form

### ISSUE THREAD

**Issue by unional** [justland/just-web#115]
This looks like a bug in TypeScript (4.8.4):

```ts
it('can detect Params type with NeedContext', () => {
  const plugin = definePlugin((options?: { a: number }) => ({
    name: 'dummy',
    init: (_: { a: number }) => ([{ dummy: options?.a }])
  }))

  // this is working
  type P = Parameters<typeof plugin>  // [options?: { a: number } | undefined]
  isType.equal<true, [options?: { a: number }], P>()
})

it('can detect Params type with NeedContext in function form', () => {
  const plugin = definePlugin((options?: { a: number }) => ({
    name: 'dummy',
    init(_: { a: number }) { return [{ dummy: options?.a }] }
  }))

  // this fails
  type P = Parameters<typeof plugin>  // []
  isType.equal<true, [options?: { a: number }], P>()
})
```

**Comment by unional** [comment]
Maybe related to https://github.com/microsoft/TypeScript/issues/43948

**Comment by unional** [comment]
It is due to this. Fixed in TS 5.1
https://github.com/microsoft/TypeScript/issues/53792
